core/src/main/java/hudson/slaves/NodeProvisioner.java

```diff
@@ -249,7 +246,29 @@ public void run() {
                                 LOGGER.log(Level.SEVERE, "Unexpected uncaught exception encountered while " 
                                         + "processing provisioned slave " + f.displayName, e);
                             } finally {
-                                itr.remove();
+                                while (true) {
+                                    List<PlannedNode> orig = pendingLaunches.get();
+                                    List<PlannedNode> repl = new ArrayList<PlannedNode>(orig);
+                                    // the contract for List.remove(o) is that the first element i where
+                                    // (o==null ? get(i)==null : o.equals(get(i)))
+                                    // is true will be removed from the list
+                                    // since PlannedNode.equals(o) is not final and we cannot assume
+                                    // that subclasses do not override with an equals which does not
+                                    // assure object identity comparison, we need to manually
+                                    // do the removal based on instance identity not equality
+                                    boolean changed = false;
+                                    for (Iterator<PlannedNode> iterator = repl.iterator(); iterator.hasNext(); ) {
+                                        PlannedNode p = iterator.next();
+                                        if (p == f) {
+                                            iterator.remove();
+                                            changed = true;
+                                            break;
+                                        }
+                                    }
+                                    if (!changed || pendingLaunches.compareAndSet(orig, repl)) {
+                                        break;
```

Comment:
so if we are not changed - or someone else has changed the list we will end up calling the cloud listener multiple times with onComplete which is a :bug: as far as I am concerned.  You can not guarantee that a Listener will not make (IMHO) correct assumptions that it will only be called once for a given event.